tools/clang/tools/dxcompiler/dxcpdbutils.cpp

```diff
@@ -85,34 +92,45 @@ static bool IsBitcode(const void *ptr, size_t size) {
   return !memcmp(ptr, pattern, _countof(pattern));
 }
 
-static bool ShouldIncludeInFlags(const std::wstring &str, bool *skip_next_arg) {
-  *skip_next_arg = false;
-  const wchar_t *specialCases[] = { L"/T", L"-T", L"-D", L"/D", L"-E", L"/E", };
-  for (unsigned i = 0; i < _countof(specialCases); i++) {
-    if (str == specialCases[i]) {
-      *skip_next_arg = true;
-      return false;
+static void ComputeFlagsBasedOnArgs(ArrayRef<std::wstring> args, std::vector<std::wstring> *outFlags, std::vector<std::wstring> *outDefines, std::wstring *outTargetProfile, std::wstring *outEntryPoint) {
+  const llvm::opt::OptTable *optionTable = hlsl::options::getHlslOptTable();
+  assert(optionTable);
+  if (optionTable) {
+    std::vector<std::string> argUtf8List;
+    for (unsigned i = 0; i < args.size(); i++) {
+      argUtf8List.push_back(ToUtf8String(args[i]));
     }
-    else if (str.find(specialCases[i]) == 0) {
-      return false;
+
+    std::vector<const char *> argPointerList;
+    for (unsigned i = 0; i < argUtf8List.size(); i++) {
+      argPointerList.push_back(argUtf8List[i].c_str());
     }
-  }
-  return true;
-}
 
-static std::vector<std::wstring> ComputeFlagsBasedOnArgs(ArrayRef<std::wstring> args) {
-  std::vector<std::wstring> flags;
-  // Flags - which exclude entry point, target profile, and defines
-  for (unsigned i = 0; i < args.size(); i++) {
-    const std::wstring &arg = args[i];
-    bool skip_another_arg = false;
-    if (ShouldIncludeInFlags(arg, &skip_another_arg)) {
-      flags.push_back(arg);
+    unsigned missingIndex = 0;
+    unsigned missingCount = 0;
+    llvm::opt::InputArgList argList = optionTable->ParseArgs(argPointerList, missingIndex, missingCount);
+    for (llvm::opt::Arg *arg : argList) {
+      if (arg->getOption().matches(hlsl::options::OPT_D)) {
+        std::wstring def = ToWstring(arg->getValue());
+        if (outDefines)
+          outDefines->push_back(def);
+        continue;
+      }
+      else if (arg->getOption().matches(hlsl::options::OPT_target_profile)) {
+        if (outTargetProfile)
+          *outTargetProfile = ToWstring(arg->getValue());
+        continue;
+      }
+      else if (arg->getOption().matches(hlsl::options::OPT_entrypoint)) {
+        if (outEntryPoint)
+          *outEntryPoint = ToWstring(arg->getValue());
+        continue;
+      }
+
+      if (outFlags)
+        outFlags->push_back(ToWstring(arg->getAsString(argList)));
```

Comment:
So, outFlags contain option and value rendered in one string.  This might not work properly for separated syntax.